File: swift/Sources/DeviceAiLlm/RAG/RagAugmentor.swift
/// Injects retrieved chunks into the messages array before sending to the engine.
/// Pure function — originals untouched. Mirrors `RagAugmentor` in kotlin/llm.
public enum RagAugmentor {

    public static func augment(messages: [LlmMessage], with chunks: [RagChunk]) -> [LlmMessage] {
        guard !chunks.isEmpty else { return messages }
        let context = "\n\nContext:\n" + chunks.map(\.text).joined(separator: "\n\n---\n\n")

        if let i = messages.firstIndex(where: { $0.role == .system }) {
            var m = messages
            m[i] = LlmMessage(role: .system, content: m[i].content + context)
            return m
        }
        return [LlmMessage(role: .system, content: "You are a helpful assistant." + context)] + messages
    }
}

File: kotlin/llm/src/commonMain/kotlin/dev/deviceai/llm/rag/RagAugmentor.kt
package dev.deviceai.llm.rag

import dev.deviceai.llm.LlmGenConfig
import dev.deviceai.llm.LlmMessage
import dev.deviceai.llm.LlmRole

/**
 * Injects retrieved RAG context into the message list before it reaches the LLM engine.
 *
 * Called by each LlmCppBridge actual inside [generate] and [generateStream].
 * Returns the original list unchanged when [LlmGenConfig.ragStore] is null.
 */
internal object RagAugmentor {

    fun augment(messages: List<LlmMessage>, config: LlmGenConfig): List<LlmMessage> {
        val store = config.ragStore ?: return messages

        // Use the last user message as the retrieval query
        val query = messages.lastOrNull { it.role == LlmRole.USER }?.content
            ?: return messages

        val chunks = store.retrieve(query, config.ragTopK)
        if (chunks.isEmpty()) return messages

        val context = chunks.joinToString("\n\n---\n\n") { it.text }
        val injected = config.ragPromptTemplate.replace("{context}", context)

        // Prepend to existing system message, or insert a new one at position 0
        val systemIdx = messages.indexOfFirst { it.role == LlmRole.SYSTEM }
        return if (systemIdx >= 0) {
            messages.toMutableList().also {
                it[systemIdx] = it[systemIdx].copy(
                    content = injected + "\n\n" + it[systemIdx].content,
                )
            }
        } else {
            listOf(LlmMessage(LlmRole.SYSTEM, injected)) + messages
        }
    }
}
